### PR TITLE
fix(checkbox): add custom events

### PR DESCRIPTION
**Pull request by anubra266**
add custom events to checkbox

### CODE REVIEW

File: packages/machines/checkbox/src/checkbox.connect.ts [hunk 1]
```diff
@@ -145,5 +145,9 @@ export function connect<T extends PropTypes>(state: State, send: Send, normalize
         }
       },
     }),
+
+    on<E extends keyof EventMap>(event: E, cb: (evt: EventMap[E]["value"]) => void) {
+      return dom.listener(state.context)(event, cb)
+    },
```
Comment (anubra266) [inline packages/machines/checkbox/src/checkbox.connect.ts:151]:
```suggestion
    on: dom.listener(state.context),
```

File: packages/machines/checkbox/src/checkbox.dom.ts [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
-  getRootEl: (ctx: Ctx) => dom.getById(ctx, dom.getRootId(ctx)),
+  getRootEl: (ctx: Ctx) => {
+    const rootEl = dom.getById(ctx, dom.getRootId(ctx))
+    if (!rootEl) throw new Error("Root element does not exist")
+    return rootEl
+  },
   getInputEl: (ctx: Ctx) => dom.getById<HTMLInputElement>(ctx, dom.getInputId(ctx)),
+
+  emitter: (ctx: Ctx) => dom.createEmitter(ctx, dom.getRootEl(ctx)),
+  listener: (ctx: Ctx) => dom.createListener(dom.getRootEl(ctx)),
```
Comment (anubra266) [inline packages/machines/checkbox/src/checkbox.dom.ts:18]:
```suggestion
  emitter: (ctx: Ctx) => dom.createEmitter<EventMap>(ctx, dom.getRootEl(ctx)),
  listener: (ctx: Ctx) => dom.createListener<EventMap>(dom.getRootEl(ctx)),

File: packages/machines/checkbox/src/checkbox.types.ts [hunk 3]
```diff
@@ -106,3 +106,12 @@ export type MachineState = {
...
 export type Send = S.Send<S.AnyEventObject>
+
+type ChangeEventDetails = { id: string; checked: boolean | "indeterminate" }
+
+type ChangeEvent = {
+  type: "change"
+  value: CustomEvent<ChangeEventDetails>
+}
+
+export type EventMap = { change: ChangeEvent }
```
Comment (anubra266) [inline packages/machines/checkbox/src/checkbox.types.ts:117]:
```suggestion
export type EventMap = { change: ChangeEventDetails }

File: packages/machines/checkbox/src/checkbox.machine.ts [hunk 4]
```diff
@@ -112,7 +112,10 @@ export function machine(ctx: UserDefinedContext) {
       actions: {
         invokeOnChange(ctx, _evt, { state }) {
           const checked = state.matches("checked")
-          ctx.onChange?.({ checked: ctx.indeterminate ? "indeterminate" : checked })
+          ctx.onChange?.({ id: ctx.id, checked: ctx.indeterminate ? "indeterminate" : checked })
+
+          const emit = dom.emitter(ctx)
+          emit("change", { id: ctx.id, checked: ctx.indeterminate ? "indeterminate" : checked })
```
Comment (segunadebayo) [inline packages/machines/checkbox/src/checkbox.machine.ts:118]: Can we refactor this to share the same details object?

File: .changeset/thin-donkeys-impress.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"@zag-js/checkbox": patch
+---
+
+Dispatch custom events
```
Comment (segunadebayo) [inline .changeset/thin-donkeys-impress.md:5]: Please elaborate here. Show a code snippet of this new API and the possibilities.

File: .changeset/thin-donkeys-impress.md [hunk 6]
```diff
@@ -3,3 +3,13 @@
...
 Dispatch custom events
+
+```js
+useEffect(() => {
+  const cleanup = api.on("change", (evt) => {
+    console.log("Checkbox is: ", evt.detail.checked ? "checked" : "unchecked")
+  })
+
+  return () => cleanup()
+}, [])
```
Comment (anubra266) [inline .changeset/thin-donkeys-impress.md:14]:
```suggestion
```js
useEffect(() => {
  return () => api.on("change", (evt) => {
    console.log("Checkbox is: ", evt.detail.checked ? "checked" : "unchecked")
  })
}, [])
```

**Review by anubra266 — COMMENTED** [review]
add listener from cleanup

File: .changeset/thin-donkeys-impress.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+Dispatch custom events
+
+```js
+useEffect(() => {
+  const cleanup = api.on("change", (evt) => {
+    console.log("Checkbox is: ", evt.detail.checked ? "checked" : "unchecked")
+  })
+
+  return () => cleanup()
+}, [])
```
Comment (anubra266) [inline .changeset/thin-donkeys-impress.md:14]:
```suggestion
```js
useEffect(() => {
  return () => api.on("change", (evt) => {
    console.log("Checkbox is: ", evt.detail.checked ? "checked" : "unchecked")
  })
}, [])
```

File: packages/machines/checkbox/src/checkbox.machine.ts [hunk 8]
```diff
@@ -112,7 +112,11 @@ export function machine(ctx: UserDefinedContext) {
       actions: {
         invokeOnChange(ctx, _evt, { state }) {
           const checked = state.matches("checked")
-          ctx.onChange?.({ checked: ctx.indeterminate ? "indeterminate" : checked })
+          const details = { id: ctx.id, checked: ctx.indeterminate ? ("indeterminate" as const) : checked }
+          ctx.onChange?.(details)
+
+          const emit = dom.emitter(ctx)
+          emit("change", details)
```
Comment (anubra266) [inline packages/machines/checkbox/src/checkbox.machine.ts:119]:
```suggestion
          const details = { id: ctx.id, checked: ctx.indeterminate ? ("indeterminate" as const) : checked }
          const emit = dom.emitter(ctx)
          emit("change", details)
          ctx.onChange?.(details)
```

File: examples/next-ts/pages/checkbox.tsx [hunk 9]
```diff
@@ -2,7 +2,7 @@ import * as checkbox from "@zag-js/checkbox"
 import { mergeProps, normalizeProps, useMachine } from "@zag-js/react"
 import { checkboxControls } from "@zag-js/shared"
 import serialize from "form-serialize"
-import { useId } from "react"
+import { useEffect, useId } from "react"
```
Comment (anubra266) [inline examples/next-ts/pages/checkbox.tsx:5]:
```suggestion
import { useId } from "react"
```

**Pull request closed without merging**